### PR TITLE
Centralize tool result error handling in format with error_line helper

### PR DESCRIPTION
**Pull request by LasmarKhalifa**
Part of https://github.com/Shopify/roast/issues/919

Foundation PR for the tool result formatter implementations.

### CODE REVIEW

**Comment by LasmarKhalifa** [comment]
* **#984** <a href="https://app.graphite.com/github/pr/Shopify/roast/984?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#983** <a href="https://app.graphite.com/github/pr/Shopify/roast/983?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#960** <a href="https://app.graphite.com/github/pr/Shopify/roast/960?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#963** <a href="https://app.graphite.com/github/pr/Shopify/roast/963?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#962** <a href="https://app.graphite.com/github/pr/Shopify/roast/962?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#959** <a href="https://app.graphite.com/github/pr/Shopify/roast/959?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#958** <a href="https://app.graphite.com/github/pr/Shopify/roast/958?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#957** <a href="https://app.graphite.com/github/pr/Shopify/roast/957?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#956** <a href="https://app.graphite.com/github/pr/Shopify/roast/956?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#955** <a href="https://app.graphite.com/github/pr/Shopify/roast/955?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#954** <a href="https://app.graphite.com/github/pr/Shopify/roast/954?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#953** <a href="https://app.graphite.com/github/pr/Shopify/roast/953?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#952** <a href="https://app.graphite.com/github/pr/Shopify/roast/952?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#951** <a href="https://app.graphite.com/github/pr/Shopify/roast/951?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#949** <a href="https://app.graphite.com/github/pr/Shopify/roast/949?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#948** <a href="https://app.graphite.com/github/pr/Shopify/roast/948?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/Shopify/roast/948?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: lib/roast/cogs/agent/providers/claude/tool_result.rb [hunk 1]
```diff
@@ -39,7 +41,15 @@ def format
...
             #: () -> String
             def format_unknown
-              "UNKNOWN [#{tool_name}] #{is_error ? " ERROR" : "OK"} #{tool_use_description || ""}\n#{content}"
+              "UNKNOWN [#{tool_name}] OK #{tool_use_description}\n#{content}"
+            end
+
+            # Renders "<TOOL> ERROR <message>" with any <tool_use_error> wrapper stripped.
+            #
+            #: () -> String
+            def error_line
```
Comment (juniper-shopify) [inline lib/roast/cogs/agent/providers/claude/tool_result.rb:75]: The doc comment here is good — it explains what the method does. Since this is a foundation helper that every tool in the stack will rely on for error formatting, it would be helpful to expand it a bit:

```ruby
# Renders "<TOOL> ERROR <message>" with any <tool_use_error> wrapper stripped.
#
# Reads the instance's +content+ and +tool_name+ to produce a single-line
# error summary. Error messages are intentionally NOT truncated so the full
# diagnostic is preserved for debugging.
#
# Examples:
#   BASH ERROR File has not been read yet.
#   UNKNOWN ERROR command not found
#
#: () -> String
def error_line
```

The reason is that someone reading a downstream formatter for the first time will look at `error_line` to understand the contract — having examples makes that a 5-second task instead of a 30-second one.

File: test/roast/cogs/agent/providers/claude/tool_result_test.rb [hunk 2]
```diff
@@ -85,6 +85,48 @@ class Claude::ToolResultTest < ActiveSupport::TestCase
           assert_match(/error details/, output)
         end
 
+        test "error_line strips the tool_use_error wrapper and upcases the tool name" do
```
Comment (juniper-shopify) [inline test/roast/cogs/agent/providers/claude/tool_result_test.rb:90]: Nice coverage here. One edge case to add: what happens when `content` is `nil`? The implementation handles it via `content.to_s`, but an explicit test documents that expectation:

```ruby
test "error_line handles nil content gracefully" do
  tool_use_message = Claude::Messages::ToolUseMessage.new(
    type: :tool_use,
    hash: { name: "read", input: {} },
  )
  tool_result = Claude::ToolResult.new(
    tool_use: tool_use_message,
    content: nil,
    is_error: true,
  )

  output = tool_result.send(:error_line)

  assert_equal "READ ERROR", output
end
```

Something to think about: when `content` is nil, this produces `"READ ERROR "` (with a trailing space) because the string interpolation gives `"READ ERROR #{""}"`. If you'd prefer `"READ ERROR"` with no trailing space, a `.strip` or `.squeeze(" ")` at the end would do it — but it's a very minor cosmetic point.

**Review by juniper-shopify — CHANGES_REQUESTED** [review]
Nice foundation PR! The error-handling centralization is clean — pulling the error path into an early-return at the top of `format` means every downstream formatter only needs to think about the success case. Makes the rest of your stack cleaner.

Two small things to address:

1. The `error_line` doc comment could use a couple of examples (see inline comment).
2. A test for `error_line` when `content` is nil would round out the edge-case coverage.

And lets switch to the xml parser we had talked about

File: lib/roast/cogs/agent/providers/claude/tool_result.rb [hunk 3]
```diff
@@ -39,7 +41,15 @@ def format
...
+              message = content.to_s.gsub(%r{</?tool_use_error>}, "").strip
```
Comment (juniper-shopify) [inline lib/roast/cogs/agent/providers/claude/tool_result.rb:76]: as we had previously discussed, let's parse this with an xml parser

Comment (LasmarKhalifa) [reply lib/roast/cogs/agent/providers/claude/tool_result.rb:76]: Just leaving this here as a general log for our decision following our pairing on this comment: We will implement this PR with the current `gsub` implementation and we will prototype a few xml parsers and other ways to parse these types of xml-looking messages. Nokogiri was previously considered as the parser, but was ultimately dropped as it contains native dependencies, which we don't want to have on Roast. Potential challenges for using a parser on error messages include the fact that unescaped special characters (`&`, `<`, `>`, etc.) can raise an error or be dropped depending on the parser, which wouldn't be ideal for an error message where these characters do appear.

**Review by juniper-shopify — APPROVED** [review]

**Comment by LasmarKhalifa** [comment]
### Merge activity

* **Jun 25, 6:03 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/Shopify/roast/948).
* **Jun 25, 6:04 PM UTC**: [Graphite](https://app.graphite.com/github/pr/Shopify/roast/948) rebased this pull request as part of a merge.
* **Jun 25, 6:06 PM UTC**: [Graphite](https://app.graphite.com/github/pr/Shopify/roast/948) rebased this pull request as part of a merge.
* **Jun 25, 6:06 PM UTC**: @LasmarKhalifa merged this pull request with [Graphite](https://app.graphite.com/github/pr/Shopify/roast/948).
